service-base/src/main/java/org/eclipse/hono/service/tenant/BaseTenantService.java

```diff
@@ -0,0 +1,305 @@
+     */
+    protected void doStart(final Future<Void> startFuture) {
+        // should be overridden by subclasses
+        startFuture.complete();
+    }
+
+    private void tenantConsumer() {
+        this.tenantConsumer = this.vertx.eventBus().consumer(TenantConstants.EVENT_BUS_ADDRESS_TENANT_IN);
+        this.tenantConsumer.handler(this::processTenantMessage);
+        this.log.info("listening on event bus [address: {}] for incoming tenant CRUD messages",
+                TenantConstants.EVENT_BUS_ADDRESS_TENANT_IN);
+    }
+
+    /**
+     * Unregisters the registration message consumer from the Vert.x event bus and then invokes {@link #doStop(Future)}.
+     *
+     * @param stopFuture the future to invoke once shutdown is complete.
+     */
+    public final void stop(final Future<Void> stopFuture) {
+        this.tenantConsumer.unregister();
+        this.log.info("unregistered tenant data consumer from event bus");
+        this.doStop(stopFuture);
+    }
+
+    /**
+     * Subclasses should override this method to perform any work required before shutting down this verticle.
+     * <p>
+     * This method is invoked by {@link #stop()} as part of the verticle deployment process.
+     * </p>
+     *
+     * @param stopFuture the future to invoke once shutdown is complete.
+     */
+    protected void doStop(final Future<Void> stopFuture) {
+        // to be overridden by subclasses
+        stopFuture.complete();
+    }
+
+    /**
+     * Processes a tenant request message received via the Vertx event bus.
+     *
+     * @param tenantMsg The message.
+     */
+    public final void processTenantMessage(final Message<JsonObject> tenantMsg) {
+        try {
+            final JsonObject body = tenantMsg.body();
+            final String tenantId = body.getString(RequestResponseApiConstants.FIELD_TENANT_ID);
+            final String operation = body.getString(MessageHelper.SYS_PROPERTY_SUBJECT);
+
+            if (tenantId == null) {
+                log.debug("tenant request does not contain mandatory property [{}]",
+                        RequestResponseApiConstants.FIELD_TENANT_ID);
+                reply(tenantMsg, TenantResult.from(HttpURLConnection.HTTP_BAD_REQUEST));
+                return;
+            }
+
+            JsonObject payload;
+
+            switch (operation) {
+            case ACTION_GET:
+                log.debug("retrieving tenant [{}]", tenantId);
+                get(tenantId, result -> reply(tenantMsg, result));
+                break;
+            case ACTION_ADD:
+                if (!isValidRequestPayload(body)) {
+                    log.debug("tenant request contains invalid structure");
+                    reply(tenantMsg, TenantResult.from(HttpURLConnection.HTTP_BAD_REQUEST));
+                    break;
+                }
+                payload = getRequestPayload(body);
+                log.debug("creating tenant [{}] with data {}", tenantId, payload != null ? payload.encode() : null);
```

Comment:
IMHO we should not log the payload since we have no control over what is in there (length, problematic characters to write to the log etc).

Reply:
:+1: would even strengthen it to "we must nost log the payload". Could be lethal to do that.